File: src/main/java/com/example/tarefas/controller/TarefaController.java
package com.example.tarefas.controller;

import com.example.tarefas.model.Tarefa;
import com.example.tarefas.repository.CategoriaRepository;
import com.example.tarefas.repository.TarefaRepository;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@RestController
@RequestMapping("/tarefas")
public class TarefaController {

    private final TarefaRepository tarefaRepository;
    private final CategoriaRepository categoriaRepository;

    public TarefaController(TarefaRepository tarefaRepository, CategoriaRepository categoriaRepository) {
        this.tarefaRepository = tarefaRepository;
        this.categoriaRepository = categoriaRepository;
    }

    @GetMapping
    public List<Tarefa> listarTodas() {
        return tarefaRepository.findAll();
    }

    @GetMapping("/{id}")
    public Tarefa buscarPorId(@PathVariable Long id) {
        return tarefaRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Tarefa não encontrada"));
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public Tarefa criar(@RequestBody Tarefa tarefa) {
        if (tarefa.getCategoria() == null || tarefa.getCategoria().getId() == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Categoria é obrigatória");
        }
        
        var categoria = categoriaRepository.findById(tarefa.getCategoria().getId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "Categoria não encontrada"));
        
        tarefa.setCategoria(categoria);

        return tarefaRepository.save(tarefa);
    }

    @PutMapping("/{id}")
    public Tarefa atualizar(@PathVariable Long id, @RequestBody Tarefa tarefa) {
        return tarefaRepository.findById(id).map(t -> {
            t.setTitulo(tarefa.getTitulo());
            t.setDescricao(tarefa.getDescricao());
            t.setDataHora(tarefa.getDataHora());

            if (tarefa.getCategoria() != null && tarefa.getCategoria().getId() != null) {
                var categoria = categoriaRepository.findById(tarefa.getCategoria().getId())
                        .orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "Categoria não encontrada"));
                t.setCategoria(categoria);
            }

            return tarefaRepository.save(t);
        }).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Tarefa não encontrada"));
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deletar(@PathVariable Long id) {
        if (!tarefaRepository.existsById(id)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Tarefa não encontrada");
        }
        tarefaRepository.deleteById(id);
    }
}

File: src/main/kotlin/com/example/tarefas/controller/TarefaController.kt
package com.example.tarefas.controller

import com.example.tarefas.model.Tarefa
import com.example.tarefas.repository.CategoriaRepository
import com.example.tarefas.repository.TarefaRepository
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException

@RestController
@RequestMapping("/tarefas")
@Tag(name = "Tarefas", description = "Endpoints para gerenciamento de tarefas")
class TarefaController(
    private val tarefaRepository: TarefaRepository,
    private val categoriaRepository: CategoriaRepository
) {

    @GetMapping
    fun listarTodas(): List<Tarefa> = tarefaRepository.findAll()

    @GetMapping("/{id}")
    fun buscarPorId(@PathVariable id: Long): Tarefa =
        tarefaRepository.findById(id).orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Tarefa não encontrada") }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun criar(@Valid @RequestBody tarefa: Tarefa): Tarefa {
        if (tarefa.categoria?.id == null) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Categoria é obrigatória")
        }

        val categoria = categoriaRepository.findById(tarefa.categoria!!.id!!)
            .orElseThrow { ResponseStatusException(HttpStatus.BAD_REQUEST, "Categoria não encontrada") }

        tarefa.categoria = categoria
        return tarefaRepository.save(tarefa)
    }

    @PutMapping("/{id}")
    fun atualizar(@PathVariable id: Long, @Valid @RequestBody tarefaAtualizada: Tarefa): Tarefa {
        val existente = buscarPorId(id)

        if (tarefaAtualizada.categoria?.id == null) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Categoria é obrigatória")
        }

        val categoria = categoriaRepository.findById(tarefaAtualizada.categoria!!.id!!)
            .orElseThrow { ResponseStatusException(HttpStatus.BAD_REQUEST, "Categoria não encontrada") }

        existente.titulo = tarefaAtualizada.titulo
        existente.descricao = tarefaAtualizada.descricao
        existente.categoria = categoria

        return tarefaRepository.save(existente)
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deletar(@PathVariable id: Long) {
        if (!tarefaRepository.existsById(id)) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, "Tarefa não encontrada")
        }
        tarefaRepository.deleteById(id)
    }
}
